draft-pda-protocol.md

```diff
@@ -562,116 +594,264 @@ PA protocol. The structure of the leader share is similarly protocol specific:
-    case prio: PrioLeaderShare;
-    case hits: HitsLeaderShare;
+  select (PALeaderInputShare.proto) {
+    case prio: PrioLeaderInputShare;
+    case hits: HitsLeaderInputShare;
   };
-} PALeaderShare;
+} PALeaderInputShare;
 ~~~
 
-Note that the leader share is sent not encrypted.
+Note that only the helper's share is encrypted, and that the leader share is
+sent in plaintext.
 
 The leader responds to well-formed requests to `[leader]/upload_finish` with
 status 200 and an empty body. Malformed requests are handled as described in
 {{pa-error-common-aborts}}.
 
-[TODO: Since we're running the protocol with multiple cohorts of aggregators, the
-collector needs to decide how to pick which cohort has the "correct" output.
-This might be the cohort with the largest batch of inputs. Figure this out once
-the collection is specified.]
+## Collect {#pa-collect}
+
+~~~~
+Collector     Leader           Helper
+  |  collect 1  |                |
+  +------------->                |
+  |             |  aggregate 1   |
+  |             <---------------->
+  |             |  ...           |
+  |             |  aggregate L   |
+  |             <---------------->
+  |             |  output share  |
+  |             <---------------->
+  <-------------+                |
+  |  ...        |                |
+  |  collect N  |                |
+  +------------->                |
+  |             |  aggregate 1   |
+  |             <---------------->
+  |             |     ...        |
+  |             |  aggregate L   |
+  |             <---------------->
+  |             |  output share  |
+  |             <---------------->
+  <-------------+                |
+  v             v                v
+~~~~
+{: #pa-collect-flow title="Flow of the collect process with N collect requests
+and L aggregate requests per collect request."}
+
+[TODO: Decide how the collector's request is authenticated.]
+
+The collector interacts with the leader to produce the final aggregate output.
+This process consists of a sequence of *collect requests* issued to the leader.
+Before a request can succeed, the aggregators must have verified and aggregated
+enough reports and the leader must have obtained the helper's encrypted output
+share (see {{pa-aggregate}}). In general, the procedure by which the aggregators
+verify and aggregate reports depends on parameters carried by the collect
+request. This procedure is described below in {{pa-aggregate}}.
+
+### Collect Request
+
+The collect protocol is an iterative procedure driven by the collector and
+parameterized by a PAParam. At a high level, it proceeds as follows. First, the
+collector initializes local per-protocol state using the corresponding
+PAParam. This state object has the following member functions:
+
+- CreateRequest() -> PACollectReq: Creates a PACollectReq object, defined below,
+  that is sent to the leader to complete one iteration of the protocol.
+
+~~~
+struct {
+  PATask task;
+  Url helper;
```

Comment:
> A `PAParam` can contain _n_ `helper_urls`, but a client may only select _k_ < _n_ of them to send reports to. So how does the collector know which `helper_urls` to use? I'm guessing that collectors are meant to send a `PACollectReq` for all _n_ helpers, but if _k_ << _n_ this seems wasteful. Are we assuming that _k_ is exactly or close to _n_?

The collector knows `helper_urls` because it's now specified by `PAParam`. (Note that this is change is new to this PR.) Nominally, _k_ would be close to _n_ (and _n_ would be be around 2 or 3).

> Supposing that some helper did not participate, what does the `PACollectResp` look like? Does it return an error (which should be described in 3.x. Common abort conditions) or does it provide a `PACollectResp` with empty `leader_share` and `encrypted_helper_share`?

Good question. (This is subsumed by the OPEN ISSUE below; I added your particular question as an example.) I think the right answer is that the leader would "abort" the request as described in {{pa-error}}. An important thing to note about this case is that it's (partially) recoverable, since the collector can try again with another helper.

> What about the case where `PACollectReq.helper` does not occur in `PAParam.helper_urls`? Does that get rejected by leader as an invalid message, or treated the same as a valid helper that did not participate?

This would be another "abort" I think.

Reply:
It seems a lot of complexity right now is coming from the fact that the collector allows for multiple helpers to be used by the client, and the client to make a selection. Can we just remove that option, and have the collector force which helpers to use?

Reply:
This discussion seems regressive given that we've already decided that multiple helpers would be used to add resiliency. If we want to discuss changing course, then let's do so in an separate issue.

Reply:
Filed: https://github.com/abetterinternet/prio-documents/issues/68